pkg/azuredx/azureauth/auth.go

```diff
@@ -0,0 +1,152 @@
+	token, err := c.ServicePrincipalToken(ctx)
+	if err != nil {
+		return "", fmt.Errorf("service principal token unavailable: %w", err)
+	}
+	return "Bearer " + token, nil
+}
+
+// QueryDataAuthorization returns an HTTP Authorization-header value which
+// represents the respective request.
+func (c *ServiceCredentials) QueryDataAuthorization(req *backend.QueryDataRequest) (string, error) {
+	switch {
+	case c.OnBehalfOf:
+		return c.queryDataOnBehalfOf(req)
+
+	// TODO(pascaldekloe): case c.ClientCreds [oauthPassThru]
+
+	default:
+		backend.Logger.Debug("using service principal token for data request")
+		ctx := context.Background()
+		if c.QueryTimeout != 0 {
+			var cancel context.CancelFunc
+			ctx, cancel = context.WithTimeout(ctx, c.QueryTimeout)
+			defer cancel()
+		}
+		return c.ServicePrincipalAuthorization(ctx)
+	}
+}
+
+func (c *ServiceCredentials) queryDataOnBehalfOf(req *backend.QueryDataRequest) (string, error) {
+	user := req.PluginContext.User // do nil-check once for all
+	if user == nil {
+		return "", errors.New("non-user requests not permitted with on-behalf-of configuration")
+	}
+
+	// Azure requires an ID token (instead of an access token) for the exchange.
+	userToken, ok := req.Headers["X-ID-Token"]
+	if !ok {
+		if _, ok := req.Headers["Authorization"]; !ok {
+			return "", errors.New("system accounts are denied with on-behalf-of configuration")
+		}
+
+		backend.Logger.Error("ID token absent for data request; enable them on the Azure portal under: “App Registrations” → the respective application → “Manage” → “Authentication”", "user", user.Login)
+		return "", errors.New("ID token absent for data request")
+	}
+
+	onBehalfOfToken, err := c.onBehalfOf(userToken)
+	if err != nil {
+		backend.Logger.Error(err.Error(), "user", user.Login)
+		// Don't leak any context to the end-user.
+		return "", errors.New("on-behalf-of token exchange failed for data request")
+	}
+	backend.Logger.Debug("aquired on-behalf-of token for data request")
+
+	return "Bearer " + onBehalfOfToken, nil
+}
+
+// OnBehalfOf resolves a token which impersonates the subject of userToken.
+// UserToken has to be an ID token. See the following link for more detail.
+// https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow
+func (c *ServiceCredentials) onBehalfOf(userToken string) (onBehalfOfToken string, err error) {
+	params := make(url.Values)
+	params.Set("client_id", c.ClientID)
+	params.Set("client_secret", c.Secret)
+	params.Set("grant_type", "urn:ietf:params:oauth:grant-type:jwt-bearer")
+	params.Set("assertion", userToken)
+	params.Set("scope", c.ClusterURL+"/.default")
+	params.Set("requested_token_use", "on_behalf_of")
+
+	// https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-v2-protocols#endpoints
+	tokenURL := "https://login.microsoftonline.com/" + url.PathEscape(c.TenantID) + "/oauth2/v2.0/token"
```

Comment:
ADX supports multiple clouds, but it looks like this url is targeting Azure public cloud only. Do you know if this api is available in all clouds? If not, we might have to disable OBO for all national clouds. If it is available in all clouds, you need to resolve the host based on the clouds. See [here](https://github.com/grafana/azure-data-explorer-datasource/blob/master/pkg/azuredx/tokenprovider/token_provider.go#L56-L68) how it's done in the token provider.

Reply:
I am not sure whether this mess applies here too. The documentation form the URL in the comment above is quite clear on it's location. I will ask a contact at Microsoft to be sure…

Reply:
It did indeed.

Reply:
Not sure I follow - it this api available in all clouds?

Reply:
@pascaldekloe

Reply:
Hard to say for sure. Needs testing. This is a new feature, so we won't break anything. The worst case is that it does not work. I say collect the knowledge with bug reports in the spirit of open-source development.

Reply:
We decided that the feature is unsupported.